packages/core/saved-objects/core-saved-objects-migration-server-internal/src/kibana_migrator_utils.ts

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import type { IndexTypesMap } from '@kbn/core-saved-objects-base-server-internal';
+import type { Logger } from '@kbn/logging';
+import type { IndexMap } from './core';
+import { TypeStatus, type TypeStatusDetails } from './kibana_migrator_constants';
+
+// even though this utility class is present in @kbn/kibana-utils-plugin, we can't easily import it from Core
+// aka. one does not simply reuse code
+export class Defer<T> {
+  public resolve!: (data: T) => void;
+  public reject!: (error: any) => void;
+  public promise: Promise<any> = new Promise<any>((resolve, reject) => {
+    (this as any).resolve = resolve;
+    (this as any).reject = reject;
+  });
+}
+
+export const defer = () => new Defer<void>();
+
+export function createMultiPromiseDefer(indices: string[]): Record<string, Defer<void>> {
+  const defers: Array<Defer<void>> = new Array(indices.length).fill(true).map(defer);
+  const all = Promise.all(defers.map(({ promise }) => promise));
+  return indices.reduce<Record<string, Defer<any>>>((acc, indexName, i) => {
+    const { resolve, reject } = defers[i];
+    acc[indexName] = { resolve, reject, promise: all };
+    return acc;
+  }, {});
+}
+
+export async function getCurrentIndexTypesMap({
+  client,
+  mainIndex,
+  defaultIndexTypesMap,
+  logger,
+}: {
+  client: ElasticsearchClient;
+  mainIndex: string;
+  defaultIndexTypesMap: IndexTypesMap;
+  logger: Logger;
+}): Promise<IndexTypesMap | undefined> {
+  try {
+    // check if the main index (i.e. .kibana) exists
+    const mapping = await client.indices.getMapping({
+      index: mainIndex,
+    });
+
+    // main index exists, try to extract the indexTypesMap from _meta
+    const meta = Object.values(mapping)?.[0]?.mappings._meta;
+    return meta?.indexTypesMap ?? defaultIndexTypesMap;
```

Comment:
We could consider a distributed approach where:
* We keep a hardcoded list of all saved object indices.
* Each index holds a list of all types stored on it. `index.mapping._meta.savedObjectTypes: string[]`

Considerations:
* ⚠️ We still need a picture of the "PRE_SPLIT" state: if that information is not present in the `_meta`, how do we know where types are stored?
* Edge case 1: Imagine that in the future we relocate a SO type `T` from index `A` to index `B`, and that `T` was the last type stored in `A`.
   * ⚠️ We can never delete `A` from the hardcoded list of indices, otherwise we would not know that we have to start a migrator for it.
    * ✅ Both A and B would be involved in the relocation, so their `_meta` would **only** be updated if both succeed, which solves your concern.